Fix for the player inadvertently being removed from the Entity Tracke…

Mordenkainen opened the pull request:
…r when the chunk they were in unloads after they teleport out of it.

This changes the chunk unload to send a forced update to any players that are listed as being in the chunk. If the player is no longer in the chunk this will cause them to update their location and properly register with the chunk they are in.

This is needed because in 1.10 the order things were done during a tick was changed. Prior to this, entities got an update prior to any chunks unloading, which gave them a chance to "move out" of the chunk and into a new one. As of 1.10, chunk unloads happen prior to updating entities, so any entity still listed as being in an unloading chunk will be removed from the entity tracker.

This causes a variety of issues including players being unable to get out of beds, and invisible players.

It is possible this needs to be extended to all entities rather than just players, but player entities are the ones with the most noticeable issues due to the change in tick order.

Mordenkainen committed:
Fix for the player inadvertently being removed from the Entity Tracker when the chunk they were in unloads after they teleport out of it.

This changes the chunk unload to send a forced update to any players that are listed as being in the chunk. If the player is no longer in the chunk this will cause them to update their location and properly register with the chunk they are in.

This is needed because in 1.10 the order things were done during a tick was changed. Prior to this, entities got an update prior to any chunks unloading, which gave them a chance to "move out" of the chunk and into a new one. As of 1.10, chunk unloads happen prior to updating entities, so any entity still listed as being in an unloading chunk will be removed from the entity tracker.

This causes a variety of issues including players being unable to get out of beds, and invisible players.

It is possible this needs to be extended to all entities rather than just players, but player entities are the ones with the most noticeable issues due to the change in tick order.

williewillus commented:
there was a vanilla bug report for this, could you add it as a comment to the patch? like `// FORGE - Fix MC-?????`

bs2609 commented:
Probably [MC-92216](https://bugs.mojang.com/browse/MC-92216) / [MC-92916](https://bugs.mojang.com/browse/MC-92916) ?

williewillus commented:
MC-92916 would be the one to comment, although I'm not sure why that ticket is closed, the issue isn't fixed afaik

Mordenkainen commented:
I was going to use MC-92216 as it is still open, many of the others are duped to it and it lists all the affected versions. If it is preferable to use 92916 instead let me know.

Ah, yes, now that I look at it, 92916 is definitely the one to go with.

bs2609 commented:
Resolved as "Awaiting Response".

Might be worth posting there, as it's still an issue.

Mordenkainen committed:
Added comment that this a fix for MC-92216

That issue is marked as resolved, but reading through the notes it indicates there is a case involving unloading chunks that the fix does not deal with. This patch addresses that case.

Mordenkainen commented:
And of course I put the wrong fix ID in the commit message. It's correct in the patch though. (and frankly, either would be valid)

I also posted some info to MC-92916 as to what I believe the root cause of the issue is, and that it is still an issue.

mezz commented on patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch:
```diff
@@ -185,15 +185,32 @@
++        for (final net.minecraft.entity.player.EntityPlayer player : players)
++        {
++            field_76637_e.func_72866_a(player, false);
++        }
```
I think this patch can be much smaller:
```java
++        // FORGE - Fix for MC-92916
++        for (ClassInheritanceMultiMap<Entity> multimap : entityLists)
++            for (net.minecraft.entity.player.EntityPlayer player : multimap.getByClass(net.minecraft.entity.player.EntityPlayer.class))
++                world.updateEntityWithOptionalForce(player, false);
```

bs2609 replied:
I think it's done that way to prevent modification of the entity lists during iteration.

Mordenkainen replied:
@bs2609 Yes, it was done that way to prevent a CME as updateEntityWithOptionalForce removes the player from the map.

@mezz I'll see what I can do.

mezz reviewed (changes requested):
Looks pretty good, just need to minimize the patch size a bit.

Mordenkainen commented:
@bs2609 Yes, it was done that way to prevent a CME.

@mezz I'll see what I can do.

Mordenkainen committed:
Moved patch code to ForgeHooks.java to reduce the footprint of the changes to Chunk.java

bs2609 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1386,4 +1387,21 @@ public static void onAdvancement(EntityPlayerMP player, Advancement advancement)
+        final List<EntityPlayer> players = new ArrayList<EntityPlayer>();
+        for (final ClassInheritanceMultiMap<Entity> multimap : entityLists)
+        {
+            for(final EntityPlayer player : multimap.getByClass(EntityPlayer.class))
```
I think this loop can be condensed to:
```java
multimap.getByClass(EntityPlayer.class).forEach(players::add);
```

Similarly, the updating can be:
```java
players.forEach(player -> world.updateEntityWithOptionalForce(player, false));
```

Not sure if that makes it short enough to be patched inline, but it's worth a thought.

Mordenkainen replied:
So I guess the patch could become:

```java
// FORGE - Fix for MC-92916
final List<net.minecraft.entity.player.EntityPlayer> players = new java.util.ArrayList<net.minecraft.entity.player.EntityPlayer>();
java.util.Arrays.stream(entityLists).forEach(multimap -> multimap.getByClass(net.minecraft.entity.player.EntityPlayer.class).forEach(players::add));
players.forEach(player -> world.updateEntityWithOptionalForce(player, false));
```

Would that be preferable to a new method in ForgeHooks? I'm more than happy to go with whatever the Forge maintainers think is best.

EDIT - But, that would require Java 8, so I think it's best not to use Streams and Lambdas to keep a wider range of compatibility.

covers1624 replied:
1.12 Depends on Java8 anyway. Its fine to use them.

Mordenkainen committed:
Shrink patch and move it back to Chunk.java.patch.

Shrunk using Streams and Lambdas.

mezz commented on patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch:
```diff
@@ -185,15 +185,23 @@
+     {
++        // FORGE - Fix for MC-92916
++        final List<net.minecraft.entity.player.EntityPlayer> players = new java.util.ArrayList<net.minecraft.entity.player.EntityPlayer>();
++        java.util.Arrays.stream(field_76645_j).forEach(multimap -> multimap.func_180215_b(net.minecraft.entity.player.EntityPlayer.class).forEach(players::add));
```
Could you use `.collect(Collectors.toList())` to construct the `players` List?
That's getting nitpicky though, I think this patch is pretty good. If you have time to clean it up more that's great.

Mordenkainen replied:
I'll play with it, but I'm not sure I can.


EDIT: Yeah, I think that may be as good as I can get it unless someone has some ideas.

The code loops through a List<List<Entity>>, For each sub-list it filters for any EntityPlayers, loops through them and adds them to the players List.

Because the .forEach on the List<List<Entity>> does not return any result, I can not .collect the results of the filtering.

At least, this is what I think. Still kinda new to all the Stream stuff.

liach replied:
Just do `final List<EntityPlayer> players = Lists.newArrayList(multimap.func_180215_b(net.minecraft.entity.player.EntityPlayer.class));`

heni123321 replied:
I had a play with this and got it down to one line with         java.util.Arrays.stream(entityLists).flatMap(multimap -> java.util.stream.StreamSupport.stream(multimap.getByClass(net.minecraft.entity.player.EntityPlayer.class).spliterator(), false)).forEach(player -> world.updateEntityWithOptionalForce(player, false));
Sorry for formatting

heni123321 replied:
I had a play with this and got it down to one line with java.util.Arrays.stream(entityLists).flatMap(multimap -> java.util.stream.StreamSupport.stream(multimap.getByClass(net.minecraft.entity.player.EntityPlayer.class).spliterator(), false)).forEach(player -> world.updateEntityWithOptionalForce(player, false));
Sorry for formatting

Mordenkainen replied:
@liach That doesn't work because there are two lists to loop through. That line assumes you already have a reference to the internal list. We are working with a ```List<List<Entity>>```, not a ```List<Entity>```

@heni123321 A step in the right direction, but I think that might be subject the a CME since it mat call ```updateEntityWithOptionalForce``` while iterating the loop. Does anyone know how steams handle this? Will it not call the lambda until the iteration is done?

HenryLoenwind replied:
why would you want to condense this piece of code into a one-liner that takes any coder 10 minutes to decode in their head and compiles into something that is slower than boring old-fashioned loops?

liach replied:
I think we may use `Iterables.concat`

liach commented:
uhh, you do not even collect it in a list. end the stream with foreach instead.

mezz commented:
As stated elsewhere in the review comments, `updateEntityWithOptionalForce` removes the player from the map so we have to be careful about CMEs. I'm not sure if streams are guaranteed to work in discrete stages or not (never looked into it), so it probably makes sense to store players to an intermediate variable.

Mordenkainen committed:
Reduce patch to a single line.

Mordenkainen committed:
Move comment to the same line. Now it truly is a single line!

LexManos commented:
This has the potential to have duplication issues at the bare minimum. Due to duplicating logic per tick. But I guess it's worth a try as nothing has popped up as an issue yet.
